### PR TITLE
[SPARK-36711][PYTHON] Support multi-index in new syntax

### PR DESCRIPTION
**Pull request by dgd-contributor**
### What changes were proposed in this pull request?
Support multi-index in new syntax to specify index data type

### Why are the changes needed?
Support multi-index in new syntax to specify index data type

https://issues.apache.org/jira/browse/SPARK-36707

### Does this PR introduce _any_ user-facing change?
After this PR user can use

``` python
>>> ps.DataFrame[[int, int],[int, int]]
typing.Tuple[pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.NameType, pyspark.pandas.typedef.typehints.NameType]


>>> arrays = [[1, 1, 2], ['red', 'blue', 'red']]
>>> idx = pd.MultiIndex.from_arrays(arrays, names=('number', 'color'))
>>> pdf = pd.DataFrame([[1,2,3],[2,3,4],[4,5,6]], index=idx, columns=["a", "b", "c"])
>>> ps.DataFrame[pdf.index.dtypes, pdf.dtypes]
typing.Tuple[pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.NameType, pyspark.pandas.typedef.typehints.NameType, pyspark.pandas.typedef.typehints.NameType]


>>> ps.DataFrame[[("index", int), ("index-2", int)], [("id", int), ("A", int)]]
typing.Tuple[pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.NameType, pyspark.pandas.typedef.typehints.NameType]


>>> ps.DataFrame[zip(pdf.index.names, pdf.index.dtypes), zip(pdf.columns, pdf.dtypes)]
typing.Tuple[pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.NameType, pyspark.pandas.typedef.typehints.NameType, pyspark.pandas.typedef.typehints.NameType]

```

### How was this patch tested?
exist tests

### CODE REVIEW

**Comment by dgd-contributor** [comment]
CC @HyukjinKwon , Could you take a look when you have time? Thank you!

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:722]: this comment seems a duplicate w/ above.

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:724]: I think this shouldn't be wrapped w/ `zip`

**Review by HyukjinKwon — COMMENTED** [review]
Mind adding an e2e test too?

**Comment by HyukjinKwon** [comment]
ok to test

File: python/pyspark/pandas/typedef/typehints.py
Comment (dgd-contributor) [reply python/pyspark/pandas/typedef/typehints.py:722]: Thanks for reviewing,

above comment is ```DataFrame[pdf.index.dtype, pdf.dtypes]```
this one is ```DataFrame[pdf.index.dtypes, pdf.dtypes]```

This one is used for :
``` python
>>> arrays = [[1, 1, 2], ['red', 'blue', 'red']]
>>> idx = pd.MultiIndex.from_arrays(arrays, names=('number', 'color'))
>>> pdf = pd.DataFrame([[1,2,3],[2,3,4],[4,5,6]], index=idx, columns=["a", "b", "c"])
>>> ps.DataFrame[pdf.index.dtypes, pdf.dtypes]
typing.Tuple[pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.NameType, pyspark.pandas.typedef.typehints.NameType, pyspark.pandas.typedef.typehints.NameType]
```

File: python/pyspark/pandas/typedef/typehints.py
Comment (dgd-contributor) [reply python/pyspark/pandas/typedef/typehints.py:724]: Sorry, could you explain more?
I use this comment for :
``` python
>>> arrays = [[1, 1, 2], ['red', 'blue', 'red']]
>>> idx = pd.MultiIndex.from_arrays(arrays, names=('number', 'color'))
>>> pdf = pd.DataFrame([[1,2,3],[2,3,4],[4,5,6]], index=idx, columns=["a", "b", "c"])
>>> ps.DataFrame[zip(pdf.index.names, pdf.index.dtypes), zip(pdf.columns, pdf.dtypes)]
typing.Tuple[pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.IndexNameType, pyspark.pandas.typedef.typehints.NameType, pyspark.pandas.typedef.typehints.NameType, pyspark.pandas.typedef.typehints.NameType]
```

Comment (HyukjinKwon) [reply python/pyspark/pandas/typedef/typehints.py:724]: I meant:

```
ps.DataFrame[(pdf.index.names, pdf.index.dtypes), zip(pdf.columns, pdf.dtypes)]
```

:-).

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [reply python/pyspark/pandas/typedef/typehints.py:722]: ohh okay, its for dtype*s*

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [reply python/pyspark/pandas/typedef/typehints.py:724]: okie, sounds good to me.

File: python/pyspark/pandas/frame.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/frame.py:2573]: Maybe we should use `SPARK_INDEX_NAME_PATTERN` and if the pattern matches.

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:727]: feel free to name it `_extract_types`.

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:726]: oh, can you remove this one while we're here? I fixed this already SPARK-36708

**Comment by HyukjinKwon** [comment]
Looks pretty promising otherwise. Thanks for working on this, @dgd-contributor.

File: python/pyspark/pandas/frame.py
Comment (dgd-contributor) [reply python/pyspark/pandas/frame.py]: Thanks, updated, please take another look when you have time!

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:762]:
```
DataFrame[["id": int, "A": int], [int, int]]
```

do you mean:

```
DataFrame[[("id", int), ("A", int)], [int, int]]
```

? `DataFrame[["id": int, "A": int]...]` is an invalid syntax in Python.

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:809]:
```suggestion
```

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:805]: Shall we add some docstring? what's input and output?

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:773]: `_convert_tuples_to_zip(data_types) if _is_valid_type_tuples(data_types) else data_types` is repeated. Can we make one function instead of two `_convert_tuples_to_zip` and `_is_valid_type_tuples`

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:828]: Looks like the generalization seems a bit odds here. `_handle_list_of_types` address unnamed type holders and `_convert_slices_to_holders` address named type holders. Can we rename it properly?

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:805]: I would name it to `_to_tuple_of_params`

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:785]: I would name it something like `_contains_names` or `_is_named_params`

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:797]: `ExtensionDtype` handling seems missing here?

File: python/pyspark/pandas/typedef/typehints.py
Comment (HyukjinKwon) [inline python/pyspark/pandas/typedef/typehints.py:858]: let's remove this method. It doesn't virtually remove the duplication by dispatching on `is_index`.

File: python/pyspark/pandas/typedef/typehints.py
Comment (dgd-contributor) [reply python/pyspark/pandas/typedef/typehints.py:785]: Thanks for your review!
I'm not understand why we have ```name``` here?

Comment (HyukjinKwon) [reply python/pyspark/pandas/typedef/typehints.py:785]: I thought it like: `slice` is always a pair of a name and type

Comment (HyukjinKwon) [reply python/pyspark/pandas/typedef/typehints.py:785]: if it's not a `slice`, it's just type and uses a default type.

File: python/pyspark/pandas/typedef/typehints.py
Comment (dgd-contributor) [reply python/pyspark/pandas/typedef/typehints.py]: Updated. Please take another look when you have time.

File: python/pyspark/pandas/typedef/typehints.py
Comment (dgd-contributor) [reply python/pyspark/pandas/typedef/typehints.py]: updated

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Shall we hold on this PR a little bit, @HyukjinKwon and committers?

**Comment by HyukjinKwon** [comment]
Yeah, let's hold off for a while.

**Comment by dgd-contributor** [comment]
I create new PR https://github.com/apache/spark/pull/34112 with ```Lead-authored-by``` and ```Co-authored-by``` tags from my personal account. Could you take a look? Thanks @HyukjinKwon

**Comment by dgd-contributor** [comment]
Please take a look at this PR https://github.com/apache/spark/pull/34176

**Pull request closed without merging**
